src/lua/protobuf.lua

```diff
@@ -0,0 +1,723 @@
+            end
+            return
+        end
+        check_integer(field_def, value)
+        if value > MAX_UINT64 or value < MIN_UINT64 then
+           error(('Input data for %q field is %d and do not fit in ' ..
+               'uint_64'):format(field_def.name, tonumber(value)))
+        end
+    end,
+    encode = function(value, field_id)
+        return wireformat.encode_uint(value, field_id)
+    end,
+}
+
+scalars.bool = {
+    encode_as_packed = true,
+    validate = function(value, field_def)
+        validate_type(field_def, value, 'boolean')
+    end,
+    encode = function(value, field_id)
+        if value then
+            return wireformat.encode_uint(1, field_id)
+        else
+            return wireformat.encode_uint(0, field_id)
+        end
+    end,
+}
+
+-- }}} Scalar type definitions
+
+
+-- {{{ Encoders
+
+local function encode_repeated(protocol, field_def, data)
+    local buf = {}
+    local encode_as_packed = false
+    if type(data) ~= 'table' then
+        error('For repeated fields table data are needed')
+    end
+    validate_table_is_array(field_def, data)
+    if is_scalar(field_def) then
+        local scalar_def = scalars[field_def.type]
+        encode_as_packed = scalar_def.encode_as_packed
+    end
+    for _, value in ipairs(data) do
+        local encoded_item = encode_field(protocol, field_def, value, true)
+        if encode_as_packed then
+            encoded_item = string.sub(encoded_item, 2)
+        end
+        table.insert(buf, encoded_item)
+    end
+    if encode_as_packed then
+        return wireformat.encode_len(table.concat(buf), field_def.id)
+    else
+        return table.concat(buf)
+    end
+end
+
+local function encode_enum(value, id_by_value, field_id, field_type)
+    if type(value) ~= 'number' and id_by_value[value] == nil then
+        error(('%q is not defined in %q enum'):format(value, field_type))
+    end
+    -- According to open enums semantics unknown enum values are encoded as
+    -- numeric identifier. https://protobuf.dev/programming-guides/enum/
+    if type(value) == 'number' then
+        local field_def = {type = 'int32', id = field_id}
+        scalars['int32'].validate(value, field_def)
+        return scalars['int32'].encode(value, field_id)
+    else
+        return scalars['int32'].encode(id_by_value[value], field_id)
+    end
+end
+
+encode_field = function(protocol, field_def, value, ignore_repeated)
+    if field_def.repeated and not ignore_repeated then
+        return encode_repeated(protocol, field_def, value)
+    elseif is_scalar(field_def) then
+        local scalar_def = scalars[field_def.type]
+        scalar_def.validate(value, field_def)
+        return scalar_def.encode(value, field_def.id)
+    elseif is_enum(protocol, field_def) then
+        local enum_def = protocol[field_def.type]
+        return encode_enum(value, enum_def['id_by_value'],
+            field_def.id, field_def.type)
```

Comment:
It seems, we can pass `protocol`, `field_def` and `value` to `encode_enum()`. It takes less arguments (3 against 4) and make all the `encode_*()` functions acces the same list of arguments.

For my taste, it makes the code more regular and so it would look better.

Reply:
Thanks!

Regarding the note: it looks fine for me.